src/machine/machine_nrf.go

```diff
@@ -96,3 +97,96 @@ func handleUART0() {
 		nrf.UART0.EVENTS_RXDRDY = 0x0
 	}
 }
+
+// I2C0 is the only I2C interface on the NRF.
+var I2C0 = I2C{}
+
+// Configure is intended to setup the I2C interface.
+func (i2c I2C) Configure(config I2CConfig) {
+	// Default I2C bus speed is 100 kHz.
+	if config.Frequency == 0 {
+		config.Frequency = TWI_FREQ_100KHZ
+	}
+
+	// do config
+	nrf.P0.PIN_CNF[SCL_PIN] = (nrf.GPIO_PIN_CNF_DIR_Input << nrf.GPIO_PIN_CNF_DIR_Pos) |
+		(nrf.GPIO_PIN_CNF_INPUT_Connect << nrf.GPIO_PIN_CNF_INPUT_Pos) |
+		(nrf.GPIO_PIN_CNF_PULL_Pullup << nrf.GPIO_PIN_CNF_PULL_Pos) |
+		(nrf.GPIO_PIN_CNF_DRIVE_S0D1 << nrf.GPIO_PIN_CNF_DRIVE_Pos) |
+		(nrf.GPIO_PIN_CNF_SENSE_Disabled << nrf.GPIO_PIN_CNF_SENSE_Pos)
+
+	nrf.P0.PIN_CNF[SDA_PIN] = (nrf.GPIO_PIN_CNF_DIR_Input << nrf.GPIO_PIN_CNF_DIR_Pos) |
+		(nrf.GPIO_PIN_CNF_INPUT_Connect << nrf.GPIO_PIN_CNF_INPUT_Pos) |
+		(nrf.GPIO_PIN_CNF_PULL_Pullup << nrf.GPIO_PIN_CNF_PULL_Pos) |
+		(nrf.GPIO_PIN_CNF_DRIVE_S0D1 << nrf.GPIO_PIN_CNF_DRIVE_Pos) |
+		(nrf.GPIO_PIN_CNF_SENSE_Disabled << nrf.GPIO_PIN_CNF_SENSE_Pos)
+
+	if config.Frequency == TWI_FREQ_400KHZ {
+		nrf.TWI0.FREQUENCY = nrf.TWI_FREQUENCY_FREQUENCY_K400
+	} else {
+		nrf.TWI0.FREQUENCY = nrf.TWI_FREQUENCY_FREQUENCY_K100
+	}
+
+	nrf.TWI0.ENABLE = nrf.TWI_ENABLE_ENABLE_Enabled
+	nrf.TWI0.PSELSCL = SCL_PIN
+	nrf.TWI0.PSELSDA = SDA_PIN
+}
+
+// Start starts an I2C communication session.
+// Nothing to do here because start is handled differently for RX and TX on NRF.
+// Strictly here for interface compatibility with AVR.
+func (i2c I2C) Start() {
+}
+
+// Stop ends an I2C communication session.
+func (i2c I2C) Stop() {
+	nrf.TWI0.TASKS_STOP = 1
+	for nrf.TWI0.EVENTS_STOPPED == 0 {
+	}
+	nrf.TWI0.EVENTS_STOPPED = 0
+}
+
+// WriteTo writes a slice of data bytes to a peripheral with a specific address.
+func (i2c I2C) WriteTo(address uint8, data []byte) {
+	nrf.TWI0.ADDRESS = nrf.RegValue(address)
+	nrf.TWI0.TASKS_STARTTX = 1
+	for _, v := range data {
+		i2c.WriteByte(v)
+	}
+
+	// Approx. 300 us needed after ending write before sending stop.
+	time.Sleep(300 * time.Microsecond)
```

Comment:
That's a rather magical constant. I'm pretty sure there must be a better way than just waiting for some time (which may or may not be correct in all cases, and is probably not the most efficient way to do it).

Also, `time.Sleep()` is somewhat special. When using goroutines, it will switch to a different goroutine if one is runnable so it may not always do what you think it does.

I know that I2C is a bit strange on the nrf. I've played with it before and it was hard to get working correctly. There probably is a way with enough trial and error. For a working (but rather complex) driver implementation see https://github.com/NordicSemiconductor/nrfx/blob/master/drivers/src/nrfx_i2s.c.

Reply:
The code here https://github.com/aykevl/tinygo/pull/37/files#diff-194bcd9b39ab8cf699c207bc6242f403R181 is supposed to wait for the TX to complete. Which it does, but issuing a STOP immediately afterward locks up the I2C bus.

I discovered that the pause was needed there by accident due to some logging I had added. Afterward I found that the smallest pause that would work was the 300 us. No doubt that is somehow related to the bus speed, but it is not completely clear how, as of yet.

The nrfx library does not uses the direct registers, so it is of limited usefulness. Worth noting is that the TWIM interface did not work only the TWI. Not sure exactly why that might be, something to do with DMA that you would probably be better equipped to figure out. This PR uses TWI which works on both nrf52 and nrf51.